src/middleware/builtin.spec.ts

```diff
@@ -0,0 +1,291 @@
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+      middleware(fakeArgs);
+      await delay();
+
+      // Assert
+      async function assertions(...args: any[]): Promise<void> {
+        // Assert that there is no error
+        assert.notExists(args[0]);
+        // The following assertion(s) check behavior that is only targeted at RegExp patterns
+        if (typeof pattern !== 'string') {
+          if (dummyContext.matches !== undefined) {
+            assert.lengthOf(dummyContext.matches, 1);
+          } else {
+            assert.fail();
+          }
+        }
+      }
+      return onNextFirstCall;
+    };
+  }
+
+  function notMatchesPatternTestCase(pattern: string | RegExp, nonMatchingText: string): Mocha.AsyncFunc {
+    return async () => {
+      // Arrange
+      const dummyContext = {};
+      const fakeNext = sinon.fake();
+      const fakeArgs = {
+        message: createFakeMessageEvent(nonMatchingText),
+        context: dummyContext,
+        next: fakeNext,
+      } as unknown as MessageMiddlewareArgs;
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+      middleware(fakeArgs);
+      await delay();
+
+      // Assert
+      assert(fakeNext.notCalled);
+      assert.notProperty(dummyContext, 'matches');
+    };
+  }
+
+  function noTextMessageTestCase(pattern: string | RegExp): Mocha.AsyncFunc {
+    return async () => {
+      // Arrange
+      const dummyContext = {};
+      const fakeNext = sinon.fake();
+      const fakeArgs = {
+        message: createFakeMessageEvent([{ type: 'divider' }]),
+        context: dummyContext,
+        next: fakeNext,
+      } as unknown as MessageMiddlewareArgs;
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+      middleware(fakeArgs);
+      await delay();
+
+      // Assert
+      assert(fakeNext.notCalled);
+      assert.notProperty(dummyContext, 'matches');
+    };
+  }
+
+  describe('using a string pattern', () => {
+    const pattern = 'foo';
+    const matchingText = 'foobar';
+    const nonMatchingText = 'bar';
+    it('should initialize', initializeTestCase(pattern));
+    it('should match message events with a pattern that matches', matchesPatternTestCase(pattern, matchingText));
+    it('should filter out message events with a pattern that does not match', notMatchesPatternTestCase(
+      pattern, nonMatchingText,
+    ));
+    it('should filter out message events which do not have text (block kit)', noTextMessageTestCase(pattern));
+  });
+
+  describe('using a RegExp pattern', () => {
+    const pattern = /foo/;
+    const matchingText = 'foobar';
+    const nonMatchingText = 'bar';
+    it('should initialize', initializeTestCase(pattern));
+    it('should match message events with a pattern that matches', matchesPatternTestCase(pattern, matchingText));
```

Comment:
Isn't this testing the same thing as the string tests?

Reply:
not exactly. since the `pattern` is a RegExp, this test verifies that the middleware can match based on a `RegExp`. also, taking a look at L45, there are a couple assertions that are specific to `RegExp` patterns.